server/index.js

```diff
@@ -104,6 +109,44 @@ app.post('/api/workout/new-exercises', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.patch('/api/workout/:workoutId', (req, res, next) => {
+  const workoutId = Number(req.body.workoutId);
+  const { exercises } = req.body;
+  if (!exercises) throw new ClientError(400, 'ERROR: No exercises found.');
+  exercises.forEach(exercise => {
+    const { exerciseId, sets } = exercise;
+    const setPromises = sets.map(set => {
+      const { reps, weight, setOrder } = set;
+      const params = [reps, weight, setOrder, workoutId, exerciseId];
+      if (setOrder === 1) {
+        const updateSql = `
+        update "sets"
+        set    "reps" = $1,
+               "weight" = $2
+        where  "setOrder" = $3
+        and     "workoutId" = $4
+        and    "exerciseId" = $5
+        returning *
+        `;
+        return db.query(updateSql, params);
+      } else {
+        const addSql = `
+        insert into "sets" ("reps", "weight", "setOrder", "workoutId", "exerciseId")
+        values             ($1, $2, $3, $4, $5)
+        returning *;
+        `;
+        return db.query(addSql, params);
+      }
+    });
+    Promise.all(setPromises)
+      .then(result => {
+        const resultSets = result.rows;
+        res.status(204).json(resultSets);
+      })
+      .catch(err => next(err));
+  });
```

Comment:
This code is potentially responding twice to the same request (if there are two or more exercises). You can use `.flatMap()` instead.

```js
const exercisePromises = exercises.flatMap(exercise => {
  const { exerciseId, sets } = exercise;
  const setPromises = sets.map(...);
  return setPromises
})
Promise.all(exercisesPromises).then(...).catch(...);

```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap